src/hotspot/share/nmt/memoryFileTracker.cpp

```diff
@@ -126,9 +126,7 @@ const GrowableArrayCHeap<MemoryFileTracker::MemoryFile*, mtNMT>& MemoryFileTrack
 
 bool MemoryFileTracker::Instance::initialize(NMT_TrackingLevel tracking_level) {
   if (tracking_level == NMT_TrackingLevel::NMT_off) return true;
-  _tracker = static_cast<MemoryFileTracker*>(os::malloc(sizeof(MemoryFileTracker), mtNMT));
-  if (_tracker == nullptr) return false;
-  new (_tracker) MemoryFileTracker(tracking_level == NMT_TrackingLevel::NMT_detail);
+  new (_tracker.as()) MemoryFileTracker(tracking_level == NMT_TrackingLevel::NMT_detail);
```

Comment:
Maybe you could add an `init` function that forwards the constructor arguments, with an extra check to see if the memory has already been initialized:

```
  template<typename...As>
  void init(As&&...args) {
    assert(is_death_pattern(), "StaticArea already initialized");
    new (as()) T(std::forward<As>(args)...);
  }
```

```suggestion
  _tracker.init(tracking_level == NMT_TrackingLevel::NMT_detail);
```

Reply:
I think that's a good idea. Unfortunately, move semantics and Rvalue references are currently undecided in the style guide, so we can't write this exact code. We can still do

```c++
template<typename... As>
void init(As&... args) {
    assert(is_death_pattern(), "StaticArea already initialized");
    new (as()) T(args...);
}
```

Which is pretty good.